### PR TITLE
4.x

### PR DESCRIPTION
**Pull request by trentm**
This is a major version bump, primarily to move the min supported Node.js to v14.5.0 (from v8.6).

## Prep checklist:

- [x] merge David's two current fix PRs to "main"
- [ ] [david] sanity check a Kibana PR updating to an elastic-apm-node@4 (https://github.com/elastic/apm-agent-nodejs/issues/3556)
- [ ] [trent] For https://github.com/elastic/apm-agent-nodejs/issues/2668, sanity check that the auto selection of the latest layer on https://www.elastic.co/guide/en/apm/agent/nodejs/current/lambda.html works properly. Will it find the latest major? Or the latest release date? *Update:* It will get it wrong. https://github.com/elastic/apm-aws-lambda/pull/402 to fix.
- [x] [trent] docs -> https://github.com/elastic/apm-agent-nodejs/pull/3601
    - Perhaps take the time now to refresh the README? Its doc links are poor.
    - update "Active release branches" section of the README.md (on "main") to include 3.x and main (for 4.x), and the "Node.js versions" section of supported-technologies doc page.
    - update this bullet in CONTRIBUTING.md in "main" (and perhaps 3.x, could cherry-pick it)
        > If this is the for the latest major (currently `3.x`)
- [x] New change for "dev/4.x" PR: Drop 'nodejs12.x nodejs10.x' from Lambda layer compatible-runtimes (in ".ci/Makefile"). Sanity check if our Lambda docs need to be updated for this. https://github.com/elastic/apm-agent-nodejs/pull/3600
- [x] [trent] finish the 4.x PR (https://github.com/elastic/apm-agent-nodejs/pull/3528):
    - [x] merge latest changes from "main"
    - [x] release `"private": true` publishing guard in package.json
    - [x] [trent, david] carefully re-review it
    - [x] review the changes to package-lock.json. There is more there than I expected. Why? Perhaps cleared after a merge of latest from "main"?
    - [x] re-read the upgrade-to-v4.asciidoc guide. Perhaps get Brandon's team to review it?
    - [x] there is at least one XXX (for the CHANGELOG4.asciidoc merge). Another in test/config.test.js.
    - [x] add explicit "--latest" flag to `gh release create ...` per https://github.com/elastic/apm-agent-nodejs/issues/2668#issuecomment-1651851442
    - [x] add explicit "--tag=latest" to npm publish in release script
    - [x] author a good commit message for this merge: https://github.com/elastic/apm-agent-nodejs/pull/3528#issuecomment-1701939332
    - [x] take it out of draft
- [ ] [trent] update docs.git for the new 4.x release docs version -> https://github.com/elastic/docs/pull/2750

## Ready to go steps/checklist (to be done in order):

- [x] update "3.x" to the current HEAD commit on "main"
- [ ] merge the "dev/4.x" PR to main
- [ ] Do a first pass attempt at updating deps. See https://github.com/elastic/apm-agent-nodejs/issues/2760 for the full list.  Minimally it would be nice to update `async-cache` and `string-similarity` to avoid the "deprecated" warnings on `npm install elastic-apm-node`.
- [ ] (wait for Monday)
- [ ] make a "4.0.0" release PR and publish 4.0.0
- [ ] (ideally we would have a new release of the Lambda ext before doing a 3.x release) 
- [ ] Do a "3.49.2" PR targetting the "3.x" branch. Be **very** sure to follow the "If this is a new major release, then:" steps in CONTRIBUTING.md -- the set `--tag=latest-3` and `--latest=false` appropriately in the release steps. (Note: I want to do a 3.x release *after* the 4.0.0 release to see if the npm dist-tag stuff works properly. 4.0.0 should still be "latest" after this release.)
- [ ] Open an issue on the robots to add branch protection for the "3.x" branch (because releases will be done from it now).

### CODE REVIEW

**Comment by trentm** [comment]
Commit message:

```
v4.x

Significant and breaking changes:
- Min supported Node.js is now 14.5.0 (up from 8.6).
- Drop support for the obsolete "patch" context manager, i.e. the
  `contextManager: "patch"` config option. This was a limited async context
  management that predated the preferred `AsyncLocalStorage` core Node.js
  mechanism for context tracking. (#3529)
- Config vars:
    - Remove `logUncaughtExceptions` config option, if the agent's
      `uncaughtException` handler is active it now *always* logs the error to
      the console. (#2412)
    - Remove `filterHttpHeaders` config option, see `sanitizeFieldNames`
      instead. (#3332)
    - Remove long deprecate `ELASTIC_APM_KUBERNETES_*` envvars, use
      `KUBERNETES_*` instead. (#2661)
    - The `useElasticTraceparentHeader` config option now defaults to `false`.
      This means the `elastic-apm-traceparent` HTTP header is now no longer
      sent by default. (#3555)
- Instrumentations:
    - Drop instrumentation for old `hapi`, the current `@hapi/hapi` is still
      instrumented. (#2691)
- APIs:
    - Ignore a `timer` option passed to `startTransaction()` and `startSpan()`
      APIs. (#2990)
    - Remove the deprecated `span.toString()` and `transaction.toString()` APIs.
      (#2348)
    - Change `apm.startTransaction()` API to return a noop transaction instead
      of null when the agent is not started. (#2429)
    - Remove `transaction.subtype` and `transaction.action` properties from the
      API. This also impacts <<apm-start-transaction>> and
      `transaction.setType(...)`, both of which now no longer accept `subtype`
      and `action` parameters. (#3557)
```

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -14,6 +14,7 @@ on:
   pull_request:
     branches:
     - main
+    - dev/4.x
```
Comment (david-luna) [inline .github/workflows/test.yml:17]: do you think it's necessary to add `3.x` here to run tests when doing maintenance PRs for bug fixing and such?

**Review by david-luna — APPROVED** [review]

File: .github/workflows/test.yml (hunk 1 above)
Comment (trentm) [reply .github/workflows/test.yml:17]: Yes, great point. I'll add that. And I guess we should remove dev/4.x before merging.

Comment (trentm) [reply .github/workflows/test.yml]: Added in commit 5ae45889.
(I'm not actually sure if instead we'll need this change to test.yml on the *3.x* branch for it to be used for pushes and PR targetting 3.x. We shall see.)

**Comment by trentm** [comment]
> * update "3.x" to the current HEAD commit on "main"

```
% git branch -f 3.x main && git push origin 3.x
Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
remote:
remote: GitHub found 3 vulnerabilities on elastic/apm-agent-nodejs's default branch (2 moderate, 1 low). To find out more, visit:
remote:      https://github.com/elastic/apm-agent-nodejs/security/dependabot
remote:
To github.com:elastic/apm-agent-nodejs.git
   f9f7ff21..7f8fd483  3.x -> 3.x
```

**Comment by trentm** [comment]
https://github.com/elastic/apm-aws-lambda/pull/407 is required to get the https://github.com/elastic/apm-aws-lambda/pull/402 fix on to the "8.9" branch, so that our apm-agent-nodejs 4.x docs build uses it. This will allow having a https://www.elastic.co/guide/en/apm/agent/nodejs/current/lambda.html doc page that uses the correct "latest version" Lambda layer, even if that is not the "latest in time" release (e.g. it'll point to a 4.0.0 release, even if there is a 3.50.0 release that comes later; or it'll point to 3.49.1 even if I do a later 4.0.0-rc.1 pre-release).

**Pull request merged**